File: Java/leetcode_1482_MinimumNumberofDaystoMakemBouquets.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        if ((long)m * k > bloomDay.length) {
            return -1;
        }
        int low = 1, high = (int) 1e9;

        while (low < high) {
            int mid = (low + high) / 2;

            if (boquets(bloomDay, m, k, mid)) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return low;      
    }
     private Boolean boquets(int[] bloomDay, int m, int k, int day) {
        int flowers = 0, runs = 0;
        for (int i = 0; i < bloomDay.length; i++) {
            if (bloomDay[i] <= day) {
                flowers++;
                if (flowers == k) {
                    runs++;
                    flowers = 0;
                }
            } else {
                flowers = 0;
            }
            if (runs == m) {
                return true;
            }
        }
        return false;
    }
}

File: Scala/leetcode_1482_MinimumNumberofDaystoMakemBouquets.scala
object Solution {
    def minDays(bloomDay: Array[Int], m: Int, k: Int): Int = {
        var boundryConst = 0
        if (m*k> bloomDay.length || m> bloomDay.length || k> bloomDay.length) {
            boundryConst= -1
            boundryConst
        } else {
            def boquets (days: Int): Boolean = {
                var flowers = 0
                var runs = 0
                var retTr = false
                var retFal = true
                for (d <- bloomDay) {
                    if  (d <= days) {
                        flowers += 1
                    } else {
                        flowers=0
                    }
                    if (flowers >= k) {
                        runs += 1
                        flowers = 0
                    }
                }
                runs >= m
            }
            var low = 1
            var high = bloomDay.max

            while (low<high) {
                var mid: Int = (high + low)/2

                if (boquets(mid)){
                    high= mid
                } else {
                    low = mid + 1
                }
            }
            low
        }
    }
}